src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+
+from .regex import (sentenceends, lastsentenceends, re_compile)
+
+# termcheck
+# buildtermdata
+# dupecheck
+
+
+
+# --------------------------------------------------------------------
+# Helper functions
+# --------------------------------------------------------------------
+def findtagreplacement(text):
+    """Search through a token to see whether it contains a replaced tag.
+    Allows finding a single tag replacement only.
+
+    text - str(): a token
+    """
+
+    tagfound = False
+    tagtype = None
+    tokens = 1
+
+    tagreplacement = re_compile(r'##@(\w+)-(\d+)##')
+    tagsreplaced = tagreplacement.search(text)
+
+    if tagsreplaced:
+        tagfound = True
+        tagtype = str(tagsreplaced.group(1))
+        tokens = int(tagsreplaced.group(2))
+
+    return (tagfound, tagtype, tokens)
+
+
+def removepunctuation(word, start=False, end=False):
+    """Removes punctuation from beginning/end of tokens (i.e. does not
+    expect sentences!).
+
+    Parameters:
+    text  - str(), token that we want to look at
+    start  - bool(): remove punctuation from start of token?
+    apos   - bool(): remove punctuation from end of token?
+    """
+
+    if isinstance(word, list):
+        if start:
+            word = [removepunctuation(word[0], start=True)] + word[1:]
+        if end:
+            word = word[:-1] + [removepunctuation(word[-1], end=True)]
+        return word
+
+    startpunctuation = '([{"\'¡¿'
+    endpunctuation = ')]}/\\"\',:;!?.‥…‼‽⁇⁈⁉'
+
+    if start:
+        word = word.lstrip(startpunctuation)
+    if end:
+        word = word.rstrip(endpunctuation)
+    return word
+
+
+def canBeDupe(word):
+    """TAGIGNORE: character sequences that should be ignored by the duplicate
+    words check
+    Ignored contents from tags can take either of the two forms below:
+       ##@lowercase-1##
+       ##@lowercase##
+    The second form is counted as one token, the first one is counted as as many
+    tokens as the number given after the dash."""
+
+    numberignore = re_compile(r'[[{(\'"\s]*[-+]?[0-9]+(?:[.,][0-9]+)*[]})\'";:.\s]*')
+
+    return len(word) > 0 and not numberignore.match(word) and not findtagreplacement(word)[0]
+
+
+def isDupe(tokens, pos):
+    """Returns how many tokens at pos are duplicated
+       tokens = ["this", "is", "is", "a", "test"]
+       pos = 2
+       return 1"""
+    # FIXME: Find a clever way to be able to check for variants of the same
+    # word, such as: a/an, singular/plural, verb tenses. It should ideally
+    # not be hardcoded here.
+    maxlen = min(3, pos, len(tokens) - pos)
+    tokens = tokens[:]
+    for l in range(1, maxlen + 1):
+        if not canBeDupe(removepunctuation(tokens[pos + l - 1], start=True, end=True)):
+            return 0
+
+        if removepunctuation(tokens[pos - l:pos], start=True) == removepunctuation(tokens[pos:pos + l], end=True):
+            return l
+
+    return 0
+
+
+def highlight(tokens, highlightstart, highlightend):
+    """TODO@sknorr
```

Comment:
"Takes a string and adds XML tags that lead to specified tokens being highlighted in the output file"